### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by thneves**
- Set up all game logics
- the game asks for the name of the players before it start
- implement winning and draw messages in the ened
- created option to play again

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    @spots = %w[1 2 3 4 5 6 7 8 9]
+    @game_on = true
+    puts 'Enter name for the first player'
+    name = gets.chomp
+    @player1 = Player.new(name, 'x', 1)
+    puts 'Enter name for the second player'
+    name = gets.chomp
+    @player2 = Player.new(name, 'o', 2)
+    puts @board
+  end
```
Comment (kubilaycaglayan) [inline lib/game.rb]: :bulb: No puts/print outside of the main.rb.

**Review by kubilaycaglayan — CHANGES_REQUESTED** [review]
#### Hello Team 👋
#### Status: Changes Required ♻️

Great job so far! 👏👋
The gameplay is amazing, everything is working fine.
Let's cover a few details together to make your project better 😊⏬

## General Requirements

- [x] Let's delete the .gitkeep file, we don't need it anymore 😊

- [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

💡 Check the inline comments, please ⏬

- [x] The game is not working out of the box. Please fix it.
[HINT] check `require`
- [x] [OPTIONAL] Using `require_relative` for files in your directory instead of `require` is a ruby convention.

image

## [REQUIRED] Functional Programming

### All the code in the `main.rb` file 😱 (current situation)
- [x] All game-logic related resources (classes and/or modules) are kept in the `lib` directory. ([screenshot] (should be)(https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

💡 Okay, you have two classes in the `lib/` directory but they are so thin. All the code is still running in the `main.rb` file.

I will share some tips with you, about how you can move methods to the lib. And how you can make your methods single responsible. Right now you don't have many functions. Everything is running with the interpreter flow.

### Beautify the main.rb
This is your main.rb. It consists of a few initial variables and a `main game loop`.
You can move items in the green box into another file.
And we have our main loop here. We will work on this while loop, more.
image

### Make a brand new function from this validator and call it here, instead of putting the code block here.
After you do this, instead of this thick code block, you will only call the method you created and you will use the return value from that method to go on with the loop. You can add this new method to your `Game` class.
image

### Make a brand new function from this code block
You can return a string value at the end and make the printing process here in main.rb. But you can put this method in your `Game` class.
image

### Make a brand new function from this code block
Similar to the above. You can move this into the `Game` class.
image

### Make a brand new function from this code block
You can move this into the `Player` class. You can initialize an instance variable named `@current_player`. And make the switching with your instance method.
image

... It goes on like this. There are two more blocks but I believe you get the idea.

😇 I know it won't be too easy to make these changes. Maybe you'll need to alter some of your main logic in your game. But please give attention to this. Because it is a really important practice for functional programming. And this is one of the hard requirements of the project. 

- [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [x] User attr_readers/attr_writers. Only use attr_accessors if you really need it and you must be aware that by using it you allow writing that property from outside the class. That variable can be set to anything. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [x] DRY rule should be respected. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/dry.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

### Finally, you can put that great big while loop into a function and call it to initialize the game.

## That's it, you can request another code review request after making the changes ♻️😇

With all my respect 🙏
> Kubilay Caglayan | Technical Support Engineer

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
+class Game
+  attr_accessor :spots, :winner, :draw, :board, :game_on, :first_run, :validator, :current_player
```
Comment (euqueme) [inline lib/game.rb]: - [x] these are the only attr_accessors you need please declare the others as attr_reader

` :current_player, :spots, :first_run, :validator, :game_on`

File: lib/player.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class Player
+  attr_accessor :name, :mark_spots, :current_player
```
Comment (euqueme) [inline lib/player.rb]: - [x] The only attr_accessor for this class should be :current_player the others could be attr_readers and your game still works correctly

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request reviewed by Maria Eugenia Queme P. (TSE)

### Hello @thneves and @Jhdezj :lollipop: :balloon: 

### Thank you for doing the changes requested by the previous reviewer :medal_sports: :medal_sports: your game and code have improved a lot since the previous review, I personally like you request for the player's names and you restart the game whenever they want to play again I think this makes your game more human friendly and stands out from the rest :video_game: :star: let's work on the following changes :construction: :construction_worker: :construction_worker: 

#### Behaviour

- [x] The game is letting players have empty names 
> let's validate that the players don't have the same name and also can't have empty names
image

- [x] The game is showing both winning and tie messages when the game is won with the last available position.

> The game should either have a winner or be a tie let's fix it
image

#### Implementation Details

- [x] You should use attr_readers/attr_writers. Only use attr_accessors if you really need it and must be aware that by using it they allow writing that property from outside the class. That variable can be set to anything. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

> It's not a good coding practice to use attr_accessors not only because your game could be a victim of cheating but also they take more space in memory 

#### General

- [x] Your linters seem to be turned off please submit for a code review when you have a green checkmark in your last commit.

### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done :memo: 

### Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

**Review by mcrd25 — APPROVED** [review]
# Changes Approved :heavy_check_mark: 

Hi Team :wave: 
Thanks for addressing the changes mentioned by previous reviewers!
I see no need for further review!

### Optional
- Consider Letting player know that they are Player 1 or Player 2 when they do not enter a name though! [OPTIONAL]


***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

*Happy coding 😄,*
**[--Maya Douglas](mayacdouglas.com)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

**Pull request merged**